Add containers_image_sequoia build tag to do simple signing verification using Sequoia-PGP, and add a signature/simplesequoia implementation

mtrmac opened the pull request:
- In `signature/internal/sequoia`, add Go bindings for https://github.com/ueno/podman-sequoia . (Parts of the middle C code are auto-generated by that project.)
- With the `containers_image_sequoia` build tag:
  - Implement the previously-included but stub-only `signature/simplesequoia` API for creating simple-signing signatures using Sequoia-PGP
  - Do _all_ simple signing verification using the Sequoia-PGP backend, instead of GnuPG via GPGME. _Hopefully_ that will be transparent to users as of now.
  - _Keep_ the traditional simple-signing API as is, backed by GnuPG via GPGME. That way, existing workflows that generate/import keys to GnuPG, or rely on `GNUPGHOME`, are unaffected.
- Add a new set of CI tests, which run with the `containers_image_sequoia` build tag; currently that requires Rawhide.

Based on original contribution in #2569, with many thanks to @ueno .

<s>DO NOT MERGE: Includes UNMERGED #2905 .</s>

ueno committed:
signature: add OpenPGP signing mechanism based on Sequoia

This adds a new OpenPGP signing mechanism based Sequoia-PGP[1]. As
Sequoia-PGP is written in Rust and doesn't provide a stable C FFI,
this integration includes a minimal shared library as a "point
solution".

To build, first follow the instruction in signature/sequoia/README.md
and install `libimage_sequoia.so*` into the library path, and then
build with the following command from the top-level directory:

  $ make BUILDTAGS="btrfs_noversion containers_image_sequoia"

Note also that, for testing on Fedora, crypto-policies settings might
need to be modified to explictly allow SHA-1 and 1024 bit RSA, as the
testing keys in signature/fixtures are using those legacy algorithms.

1. https://sequoia-pgp.org/

Signed-off-by: Daiki Ueno <dueno@redhat.com>

mtrmac committed:
Allow using sequoia in macOS

- Use the correct library name
- Allow specifying a library path at compile time

Example usage:
> make BUILDTAGS="btrfs_noversion containers_image_sequoia" SEQUOIA_SONAME_DIR=$(pwd)/signature/internal/sequoia/rust/target/release

Signed-off-by: Miloslav Trmač <mitr@redhat.com>

mtrmac committed:
Merge branch 'wip/signature-sequoia' of https://github.com/ueno/containers-image into signature-sequoia

mtrmac committed:
Direct Rust logging to logrus

By default, the Rust logging output is silently discarded.

Signed-off-by: Miloslav Trmač <mitr@redhat.com>

ueno commented on signature/internal/sequoia/rust/src/signature.rs:
```diff
@@ -0,0 +1,461 @@
+                                return Ok(());
+                            }
+                            Err(verification_error) => {
+                                signature_errors.push(verification_error.to_string());
```
> > I would keep the error types and let the caller format them:
> 
> Thanks, that might very well be cleaner. It would require a somewhat tight coupling with the caller — the caller would specifically need to know that it should ignore the returned “no valid signature” error.

Actually that's the intention behind this suggestion: if the errors are converted into a string at the callee, important information that could be retrieved through [VerificationError](https://docs.rs/sequoia-openpgp/latest/sequoia_openpgp/parse/stream/enum.VerificationError.html) may be lost, or it would require parsing of the error strings later.

> > I also wonder if we might want to record all signers if there are multiple signatures.
> 
> For myself, I think I need to study / understand what is the precise semantics of `MessageStructure`, whether it is possible to have multiple signatures, and what does that mean.
> 
> … and whether it is relevant to designing a stable C API, which is the ~most urgent work right now.

`MessageStructure` can certainly have multiple signatures, though the current c/image API seems to only expect a single signature to be valid. For future extension with PQ/T multiple [signatures](https://www.ietf.org/archive/id/draft-ietf-openpgp-pqc-12.html#name-multiple-signatures), it might make sense to provide a choice on which signature has to be valid and which are not, through the c/image API, or a policy for that might be sufficient.

As for the C API, GPGME provides a similar [abstraction](https://www.gnupg.org/documentation/manuals/gpgme/Verify.html) with `gpgme_verify_result_t`, which also supports multiple signatures.

mtrmac committed:
Add a ~representative test of the typical workflow to sequoia.SigningMechanism

Signed-off-by: Miloslav Trmač <mitr@redhat.com>

mtrmac commented on signature/internal/sequoia/sequoia_test.go:
```diff
@@ -0,0 +1,228 @@
+//go:build containers_image_sequoia
+
+package sequoia_test
```
This can use `package sequoia`, and potentially access private APIs.

mtrmac committed:
With sequoia, still use GPGME for existing signing, and add a new Signer API

Signed-off-by: Miloslav Trmač <mitr@redhat.com>

mtrmac committed:
Modify signature/internal/sequoia tests to run in the same package

Should not change test behavior, merely to be more similar
to the rest of the codebase.

Signed-off-by: Miloslav Trmač <mitr@redhat.com>

mtrmac committed:
Move sequoia initialization out of init()

- Allow using c/image without loading sequoia if
  GO_SEQUOIA_ENABLE_DLOPEN is used, and nothing uses
  the Sequoia code at runtime.
- Allow adding ~independent users of signature/internal/sequoia,
  without having to coordinate to call Init at least once.
  We will add signature/simplesequoia to benefit from this.

Signed-off-by: Miloslav Trmač <mitr@redhat.com>

mtrmac committed:
Allow using the default Sequoia home

We can't reasonably unit-test the $HOME variant; tested manually.

Signed-off-by: Miloslav Trmač <mitr@redhat.com>

mtrmac committed:
Test containers_image_sequoia in CI

CI images come from https://github.com/containers/automation_images/pull/411 .

Signed-off-by: Miloslav Trmač <mitr@redhat.com>

mtrmac commented on signature/internal/sequoia/rust/Cargo.toml:
```diff
@@ -0,0 +1,42 @@
+env_logger = "0.11.8"
+libc = "0.2"
+log = "0.4"
+sequoia-cert-store = { version = "0.7", default-features = false }
```
… and removing it saves 1.5 MB.

mtrmac committed:
Don't leak SequoiaMechanism instances

Signed-off-by: Miloslav Trmač <mitr@redhat.com>

mtrmac committed:
Remove SupportsSigning from sequoia.SigningMechanism

It can be handled in callers - and it's better to do based on .Init()
outcomes.

Signed-off-by: Miloslav Trmač <mitr@redhat.com>

mtrmac commented on signature/simplesequoia/signer.go:
```diff
@@ -0,0 +1,134 @@
+// WithSequoiaHome returns an Option for NewSigner, specifying a Sequoia home directory to use.
+func WithSequoiaHome(sequoiaHome string) Option {
+	return func(s *simpleSequoiaSigner) error {
+		if sequoiaHome == "none" || sequoiaHome == "default" { // FIXME: Do we need to special-case this?
```
Thanks, dropped.

ueno commented on signature/internal/sequoia/sequoia.h:
```diff
@@ -0,0 +1,85 @@
+typedef enum SequoiaErrorKind {
+  Unknown,
+  InvalidArgument,
+  IoError,
```
Maybe we could use the [`prefix-with-name`](https://github.com/mozilla/cbindgen/blob/master/docs.md#enum-annotations) annotation in cbindgen?

mtrmac replied:
Thanks!

I will update this PR to refer to that tree.

mtrmac commented on signature/internal/sequoia/rust/src/signature.rs:
```diff
@@ -402,7 +399,7 @@ pub unsafe extern "C" fn sequoia_import_result_get_content(
         return ptr::null();
     }
-    let key_handle = &(*result_ptr).key_handles[index];
+    let key_handle = &(&(*result_ptr)).key_handles[index];
```
Without this, I see (with Rust 1.88.0, even on `cargo build --release)
```
   Compiling image-sequoia v0.1.0 (…/image/signature/internal/sequoia/rust)
warning: implicit autoref creates a reference to the dereference of a raw pointer
   --> src/signature.rs:404:23
    |
404 |     let key_handle = &(*result_ptr).key_handles[index];
    |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: creating a reference requires the pointer target to be valid and imposes aliasing requirements
    = note: `#[warn(dangerous_implicit_autorefs)]` on by default
help: try using a raw pointer method instead; or if this reference is intentional, make it explicit
    |
404 |     let key_handle = &(&(*result_ptr).key_handles)[index];
    |                       ++                         +

warning: `image-sequoia` (lib) generated 1 warning
    Finished `release` profile [optimized] target(s) in 2.58s
```

and Clippy 0.1.88 does not complain about the `&(&(*` version.

Anyway, I’m perfectly happy to drop this; I wanted a no-warnings build for the changes I’m proposing, because I’m so new to Rust, but I have no opinion on best practices or relevant versions.

mtrmac committed:
Use github.com/ueno/podman-sequoia instead of a local copy of the code

Signed-off-by: Miloslav Trmač <mitr@redhat.com>

mtrmac committed:
Add missing error handling to go_sequoia_import_result_get_content .

Signed-off-by: Miloslav Trmač <mitr@redhat.com>

mtrmac committed:
Fix a memory leak when loading libpodman_sequoia

Signed-off-by: Miloslav Trmač <mitr@redhat.com>

mtrmac committed:
Close mechanisms in tests

Signed-off-by: Miloslav Trmač <mitr@redhat.com>

mtrmac committed:
Improve test coverage of signature/internal/sequoia/sequoia.go

Add missing tests, or add comments where that's not practical.

Should not change behavior.

Signed-off-by: Miloslav Trmač <mitr@redhat.com>

mtrmac committed:
Improve test coverage of signature/*_sequoia.go

Should not change behavior.

Signed-off-by: Miloslav Trmač <mitr@redhat.com>

mtrmac committed:
Update the documentation of sequoiaSigningMechanism.Verify

... to match the recent more specific subkey validation
semantics.

Should not change behavior.

Signed-off-by: Miloslav Trmač <mitr@redhat.com>

mtrmac commented on signature/internal/sequoia/gosequoia.c:
```diff
@@ -0,0 +1,200 @@
+/*
```
`signature/internal/sequoia/*.[ch]` are auto-generated in https://github.com/ueno/podman-sequoia/tree/main/go/sequoia , only copied into this repo.

lsm5 commented:
> Ready for review, PTAL. I have preserved the original work from #2569 and this is structured as a series of commits on top, but in this case, it’s probably easier to just review the final diff as a single unit.
> 
> Cc: @giuseppe @Luap99 @lsm5

Should this PR be moved to the monorepo?

Luap99 reviewed (approved):
LGTM, I rather get this merged by EOD before the monorepo move to avoid having to redo all commits against the new repo structure.

Automation wise it means we must port the rawhide tests to the new repo but we can do this after the initial repo creation.
cc @jankaluza

lsm5 reviewed (approved):
> LGTM, I rather get this merged by EOD before the monorepo move to avoid having to redo all commits against the new repo structure.

LGTM in that case
